scripts/generate/templates/Heap.js

```diff
@@ -0,0 +1,327 @@
+    ${node} storage rootData = _unsafeNodeAccess(self, rootIdx);
+
+    // cache old value and replace it
+    ${valueType} oldValue = rootData.value;
+    rootData.value = newValue;
+
+    // re-heapify
+    _siftDown(self, size, 0, newValue, comp);
+
+    // return old root value
+    return oldValue;
+}
+
+/**
+ * @dev Returns the number of elements in the heap.
+ */
+function length(${struct} storage self) internal view returns (${indexType}) {
+    return self.data.length.to${capitalize(indexType)}();
+}
+
+/**
+ * @dev Removes all elements in the heap.
+ */
+function clear(${struct} storage self) internal {
+    ${struct}Node[] storage data = self.data;
+    /// @solidity memory-safe-assembly
+    assembly {
+        sstore(data.slot, 0)
+    }
+}
+
+/*
+ * @dev Swap node \`i\` and \`j\` in the tree.
+ */
+function _swap(${struct} storage self, ${indexType} i, ${indexType} j) private {
+    ${node} storage ni = _unsafeNodeAccess(self, i);
+    ${node} storage nj = _unsafeNodeAccess(self, j);
+    ${indexType} ii = ni.index;
+    ${indexType} jj = nj.index;
+    // update pointers to the data (swap the value)
+    ni.index = jj;
+    nj.index = ii;
+    // update lookup pointers for consistency
+    _unsafeNodeAccess(self, ii).lookup = j;
+    _unsafeNodeAccess(self, jj).lookup = i;
+}
+
+/**
+ * @dev Perform heap maintenance on \`self\`, starting at position \`pos\` (with the \`value\`), using \`comp\` as a
+ * comparator, and moving toward the leafs of the underlying tree.
+ *
+ * NOTE: This is a private function that is called in a trusted context with already cached parameters. \`length\`
+ * and \`value\` could be extracted from \`self\` and \`pos\`, but that would require redundant storage read. These
+ * parameters are not verified. It is the caller role to make sure the parameters are correct.
+ */
+function _siftDown(
+    ${struct} storage self,
+    ${indexType} size,
+    ${indexType} pos,
+    ${valueType} value,
+    function(uint256, uint256) view returns (bool) comp
+) private {
+    uint256 left = 2 * pos + 1; // this could overflow ${indexType}
+    uint256 right = 2 * pos + 2; // this could overflow ${indexType}
+
+    if (right < size) {
+        // the check guarantees that \`left\` and \`right\` are both valid uint32
+        ${indexType} lIndex = ${indexType}(left);
+        ${indexType} rIndex = ${indexType}(right);
+        ${valueType} lValue = _unsafeNodeAccess(self, _unsafeNodeAccess(self, lIndex).index).value;
+        ${valueType} rValue = _unsafeNodeAccess(self, _unsafeNodeAccess(self, rIndex).index).value;
+        if (comp(lValue, value) || comp(rValue, value)) {
+            if (comp(lValue, rValue)) {
+                _swap(self, pos, lIndex);
+                _siftDown(self, size, lIndex, value, comp);
+            } else {
+                _swap(self, pos, rIndex);
+                _siftDown(self, size, rIndex, value, comp);
+            }
```

Comment:
I think is more readable to do:

```suggestion
            index = ternary(comp(lValue, rValue), lIndex, rIndex);
            _swap(self, pos, index);
            _siftDown(self, size, index, value, comp);
```

Though, I understand it may require an extra value in the stack so it's not more optimal. Just leaving the suggestion